File: pt4/1027.longest-arithmetic-subsequence.java
import java.util.HashMap;
import java.util.Map;

/*
 * 1027. Longest Arithmetic Subsequence
 */
class Solution {
    public int longestArithSeqLength(int[] nums) {
        @SuppressWarnings("unchecked")
        Map<Integer, Integer>[] dp = new HashMap[nums.length];
        for (int i = 0; i < nums.length; i++) {
            dp[i] = new HashMap<>();
        }

        int best = 2;
        for (int right = 0; right < nums.length; right++) {
            for (int left = 0; left < right; left++) {
                int diff = nums[right] - nums[left];
                int length = dp[left].getOrDefault(diff, 1) + 1;
                dp[right].put(diff, Math.max(dp[right].getOrDefault(diff, 1), length));
                best = Math.max(best, dp[right].get(diff));
            }
        }

        return best;
    }
}

/*
Interview Explanation

Core idea:
An arithmetic subsequence is defined by its last index and common difference.
If a sequence ending at left has difference d, nums[right] can extend it when
nums[right] - nums[left] == d.

Java data structures:
- Map<Integer, Integer>[] dp stores, for every index, the best length for each
  difference ending at that index.
- HashMap is appropriate because differences can be negative and sparse.

Algorithm:
For every pair left < right:
1. Compute diff = nums[right] - nums[left].
2. Extend dp[left][diff] by one, defaulting to length 1 before adding nums[right].
3. Store the best length in dp[right][diff].
4. Track the global maximum.

Correctness:
Every arithmetic subsequence of length at least two has a final pair
(left, right) and a common difference diff. The transition extends the optimal
subsequence ending at left with that diff. Since all possible final pairs are
processed, the best recorded length is the longest arithmetic subsequence.

Complexity:
There are O(n^2) pairs and HashMap operations are O(1) average, so time is
O(n^2). In the worst case, each pair creates a distinct difference, so space is
O(n^2).

Edge cases:
- All values equal use diff 0.
- Decreasing sequences use negative differences.
- Minimum length 2 returns 2.
*/

File: pt4/1027.longest-arithmetic-subsequence.go
package main

/*
1027. Longest Arithmetic Subsequence
*/
func longestArithSeqLength(nums []int) int {
	dp := make([]map[int]int, len(nums))
	for i := range dp {
		dp[i] = make(map[int]int)
	}

	best := 2
	for right := 0; right < len(nums); right++ {
		for left := 0; left < right; left++ {
			diff := nums[right] - nums[left]
			length := dp[left][diff] + 1
			if dp[left][diff] == 0 {
				length = 2
			}
			if length > dp[right][diff] {
				dp[right][diff] = length
			}
			if dp[right][diff] > best {
				best = dp[right][diff]
			}
		}
	}

	return best
}

/*
Interview Explanation

Core idea:
An arithmetic subsequence is determined by its last index and common
difference. If a sequence ending at left has difference d, nums[right] can
extend it when nums[right] - nums[left] == d.

Go data structures:
- []map[int]int stores a map for each ending index.
- The map key is the difference, and the value is the best length ending at
  that index with that difference. A map is natural because differences can be
  negative and sparse.

Algorithm:
For every pair left < right:
1. Compute diff.
2. Extend the best sequence ending at left with that diff, or start a length-2
   sequence if none exists.
3. Store the best length in dp[right][diff].
4. Track the global maximum.

Correctness:
Every arithmetic subsequence of length at least two has a final pair and a
common difference. The transition considers that final pair and extends the
best valid earlier sequence. Since all pairs are processed, every possible
subsequence ending is represented, and the maximum is the answer.

Complexity:
There are O(n^2) pairs and average O(1) map operations, so time is O(n^2).
Space is O(n^2) in the worst case.

Edge cases:
- Equal values use diff 0.
- Decreasing sequences use negative differences.
- Minimum input length 2 returns 2.
*/
